composer/datasets/hf_streaming.py

```diff
@@ -0,0 +1,243 @@
+from dataclasses import dataclass
+from functools import partial
+from itertools import chain
+from typing import List, Optional
+
+import yahp as hp
+from torch.utils.data import IterableDataset
+
+from composer.core.types import Batch, DataSpec
+from composer.datasets.dataloader import DataloaderHparams
+from composer.datasets.hparams import DatasetHparams
+from composer.utils import dist
+
+log = logging.getLogger(__name__)
+
+
+def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
+    if isinstance(batch, dict):
+        chunked = {k: v.chunk(n_microbatches) for k, v in batch.items()}
+        num_chunks = len(list(chunked.values())[0])
+        return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
+    else:
+        raise ValueError(f'Expect batch from dataloader to be of type Dict[str, Tensor], but got {type(batch)}')
+
+
+CACHED_DATASET_SIZES = {"c4": {"en": {"train": (1024, 356317), "validation": (8, 45576)}}}
+
+
+@dataclass
+class HFStreamingDatasetHparams(DatasetHparams):
+    """Defines a generic dataset class for autoregressive and masked language models."""
+
+    name: str = hp.optional("Name of the dataset to load.", default=None)
+    config_name: Optional[str] = hp.optional(
+        "If required, the specific configuration of the dataset that you would like to use.", default=None)
+    split: str = hp.optional("What split of the dataset to use (e.g. 'train' or 'validation' or 'test')", default=None)
+    max_shards: int = hp.optional("Max number of shards, used to deterministically reduce dataset size.", default=-1)
+    max_samples: int = hp.optional(
+        "Max number of post-processed samples, note that the subset will depend on seed and world size.", default=-1)
+    tokenizer_name: str = hp.optional("The name of the tokenizer to preprocess text with.", default=None)
+    max_seq_len: int = hp.optional("The max sequence length of each token sample.", default=None)
+    group_method: str = hp.optional("How to group text samples into token samples.", default=None)
+    use_masked_lm: bool = hp.optional("Whether the dataset shoud be encoded with masked language modeling or not.",
+                                      default=None)
+    mlm_probability: float = hp.optional("If using masked language modeling, the probability to mask tokens with.",
+                                         default=0.15)
+    shuffle_buffer_size: int = hp.optional("How many incoming data samples should be kept in a buffer to shuffle.",
+                                           default=10000)
+    seed: int = hp.optional("Which seed to use to generate train and validation splits.", default=5)
+    shuffle: bool = hp.optional("Whether to shuffle the dataset for each epoch.", default=True)
+    drop_last: bool = hp.optional("Whether to drop the last samples for the last batch.", default=False)
+
+    def validate(self):
+        assert self.name is not None, "Must provide a dataset name."
```

Comment:
For validation, can you using ValueErrors? assert should not be used for data validation (see https://github.com/mosaicml/composer/blob/dev/STYLE_GUIDE.md#4-use-of-assert)